"Cannot get node id for DirectoryArtifactValue" on tree artifact containing symlink + remote cache + BES

Repro:

defs.bzl
```
def _impl(ctx):
    d = ctx.actions.declare_directory(ctx.label.name)
    ctx.actions.run_shell(
        outputs = [d],
        command = "cd $1 && mkdir dir && touch dir/file.txt && ln -s dir sym",
        arguments = [d.path],
    )
    return DefaultInfo(files = depset([d]))

buggy = rule(_impl)
```

BUILD
```
load("//:defs.bzl", "buggy")

buggy(
    name = "buggy",
)
```

.bazelrc
```
build --remote_instance_name=projects/bazel-untrusted/instances/default_instance
build --remote_cache=grpcs://remotebuildexecution.googleapis.com
build --remote_download_toplevel
build --google_default_credentials
build --bes_backend=buildeventservice.googleapis.com
build --bes_instance_name=bazel-untrusted
build --bes_results_url=https://source.cloud.google.com/results/invocations
```

Build once to populate the remote cache, then clean and rebuild. The following stack trace is produced:

```
java.lang.RuntimeException: Unexpected Exception 'Cannot get node id for DirectoryArtifactValue{mtime=1701528378671}' when closing BEP transports, this is a bug.
	at com.google.devtools.build.lib.buildeventservice.BuildEventServiceModule.waitForBuildEventTransportsToClose(BuildEventServiceModule.java:543)
	at com.google.devtools.build.lib.buildeventservice.BuildEventServiceModule.closeBepTransports(BuildEventServiceModule.java:625)
	at com.google.devtools.build.lib.buildeventservice.BuildEventServiceModule.afterCommand(BuildEventServiceModule.java:640)
	at com.google.devtools.build.lib.runtime.BlazeRuntime.afterCommand(BlazeRuntime.java:627)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:688)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:244)
	at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:550)
	at com.google.devtools.build.lib.server.GrpcServerImpl.lambda$run$1(GrpcServerImpl.java:621)
	at io.grpc.Context$1.run(Context.java:566)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.util.concurrent.ExecutionException: java.lang.UnsupportedOperationException: Cannot get node id for DirectoryArtifactValue{mtime=1701528378671}
	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:592)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:571)
	at com.google.common.util.concurrent.AbstractFuture$TrustedFuture.get(AbstractFuture.java:111)
	at com.google.common.util.concurrent.Uninterruptibles.getUninterruptibly(Uninterruptibles.java:247)
	at com.google.devtools.build.lib.buildeventservice.BuildEventServiceModule.waitForBuildEventTransportsToClose(BuildEventServiceModule.java:529)
	... 11 more
Caused by: java.lang.UnsupportedOperationException: Cannot get node id for DirectoryArtifactValue{mtime=1701528378671}
	at com.google.devtools.build.lib.remote.RemoteActionFileSystem$2.getNodeId(RemoteActionFileSystem.java:665)
	at com.google.devtools.build.lib.vfs.DigestUtils$CacheKey.<init>(DigestUtils.java:67)
	at com.google.devtools.build.lib.vfs.DigestUtils.manuallyComputeDigest(DigestUtils.java:193)
	at com.google.devtools.build.lib.vfs.DigestUtils.getDigestWithManualFallback(DigestUtils.java:160)
	at com.google.devtools.build.lib.remote.util.DigestUtil.compute(DigestUtil.java:72)
	at com.google.devtools.build.lib.remote.util.DigestUtil.compute(DigestUtil.java:67)
	at com.google.devtools.build.lib.remote.ByteStreamBuildEventArtifactUploader.readPathMetadata(ByteStreamBuildEventArtifactUploader.java:239)
	at com.google.devtools.build.lib.remote.ByteStreamBuildEventArtifactUploader.lambda$doUpload$8(ByteStreamBuildEventArtifactUploader.java:413)
	at io.reactivex.rxjava3.internal.operators.flowable.FlowableMap$MapSubscriber.onNext(FlowableMap.java:64)
	at io.reactivex.rxjava3.internal.operators.flowable.FlowableFromIterable$IteratorSubscription.fastPath(FlowableFromIterable.java:185)
	at io.reactivex.rxjava3.internal.operators.flowable.FlowableFromIterable$BaseRangeSubscription.request(FlowableFromIterable.java:129)
	at io.reactivex.rxjava3.internal.subscribers.BasicFuseableSubscriber.request(BasicFuseableSubscriber.java:153)
	at io.reactivex.rxjava3.internal.jdk8.FlowableCollectWithCollectorSingle$CollectorSingleObserver.onSubscribe(FlowableCollectWithCollectorSingle.java:102)
	at io.reactivex.rxjava3.internal.subscribers.BasicFuseableSubscriber.onSubscribe(BasicFuseableSubscriber.java:67)
	at io.reactivex.rxjava3.internal.operators.flowable.FlowableFromIterable.subscribe(FlowableFromIterable.java:69)
	at io.reactivex.rxjava3.internal.operators.flowable.FlowableFromIterable.subscribeActual(FlowableFromIterable.java:47)
	at io.reactivex.rxjava3.core.Flowable.subscribe(Flowable.java:15917)
	at io.reactivex.rxjava3.internal.operators.flowable.FlowableMap.subscribeActual(FlowableMap.java:38)
	at io.reactivex.rxjava3.core.Flowable.subscribe(Flowable.java:15917)
	at io.reactivex.rxjava3.internal.jdk8.FlowableCollectWithCollectorSingle.subscribeActual(FlowableCollectWithCollectorSingle.java:71)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap.subscribeActual(SingleFlatMap.java:37)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap.subscribeActual(SingleFlatMap.java:37)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap.subscribeActual(SingleFlatMap.java:37)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleUsing.subscribeActual(SingleUsing.java:83)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
	at io.reactivex.rxjava3.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)
	at io.reactivex.rxjava3.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:25)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	... 3 more
```

Reproducible on 7.0.0rc5. On 6.4.0 it fails due to a different error, which can be fixed with https://github.com/bazelbuild/bazel/pull/20409. See original report at https://github.com/bazelbuild/bazel/issues/20408.

--- a/src/main/java/com/google/devtools/build/lib/actions/ActionExecutedEvent.java
+++ b/src/main/java/com/google/devtools/build/lib/actions/ActionExecutedEvent.java
@@ -173,7 +173,7 @@ public final class ActionExecutedEvent implements BuildEventWithConfiguration {
       localFiles.add(
           new LocalFile(
               primaryOutput,
-              LocalFileType.forArtifact(outputArtifact),
+              LocalFileType.forArtifact(outputArtifact, primaryOutputMetadata),
               outputArtifact,
               primaryOutputMetadata));
     }

--- a/src/main/java/com/google/devtools/build/lib/analysis/TargetCompleteEvent.java
+++ b/src/main/java/com/google/devtools/build/lib/analysis/TargetCompleteEvent.java
@@ -406,12 +406,13 @@ public final class TargetCompleteEvent
             new ArtifactReceiver() {
               @Override
               public void accept(Artifact artifact) {
+                FileArtifactValue metadata = completionContext.getFileArtifactValue(artifact);
                 builder.add(
                     new LocalFile(
                         completionContext.pathResolver().toPath(artifact),
-                        LocalFileType.forArtifact(artifact),
+                        LocalFileType.forArtifact(artifact, metadata),
                         artifact,
-                        completionContext.getFileArtifactValue(artifact)));
+                        metadata));
               }
 
               @Override

--- a/src/main/java/com/google/devtools/build/lib/buildeventstream/BuildEvent.java
+++ b/src/main/java/com/google/devtools/build/lib/buildeventstream/BuildEvent.java
@@ -66,7 +66,22 @@ public interface BuildEvent extends ChainableEvent, ExtendedEventHandler.Postabl
             || this == OUTPUT_SYMLINK;
       }
 
-      public static LocalFileType forArtifact(Artifact artifact) {
+      /**
+       * Returns the {@link LocalFileType} implied by a {@link FileArtifactValue}, or the associated
+       * {@link Artifact} if metadata is not available.
+       */
+      public static LocalFileType forArtifact(
+          Artifact artifact, @Nullable FileArtifactValue metadata) {
+        if (metadata != null) {
+          switch (metadata.getType()) {
+            case DIRECTORY:
+              return LocalFileType.OUTPUT_DIRECTORY;
+            case SYMLINK:
+              return LocalFileType.OUTPUT_SYMLINK;
+            default:
+              return LocalFileType.OUTPUT_FILE;
+          }
+        }
         if (artifact.isDirectory()) {
           return LocalFileType.OUTPUT_DIRECTORY;
         } else if (artifact.isSymlink()) {

--- a/src/main/java/com/google/devtools/build/lib/runtime/NamedArtifactGroup.java
+++ b/src/main/java/com/google/devtools/build/lib/runtime/NamedArtifactGroup.java
@@ -75,14 +75,14 @@ class NamedArtifactGroup implements BuildEvent {
     for (Object elem : set.getLeaves()) {
       ExpandedArtifact expandedArtifact = (ExpandedArtifact) elem;
       if (expandedArtifact.relPath == null) {
-        FileArtifactValue fileMetadata =
+        FileArtifactValue metadata =
             completionContext.getFileArtifactValue(expandedArtifact.artifact);
         artifacts.add(
             new LocalFile(
                 completionContext.pathResolver().toPath(expandedArtifact.artifact),
-                getOutputType(fileMetadata),
-                fileMetadata == null ? null : expandedArtifact.artifact,
-                fileMetadata));
+                LocalFileType.forArtifact(expandedArtifact.artifact, metadata),
+                metadata == null ? null : expandedArtifact.artifact,
+                metadata));
       } else {
         // TODO(b/199940216): Can fileset metadata be properly handled here?
         artifacts.add(
@@ -94,20 +94,6 @@ class NamedArtifactGroup implements BuildEvent {
       }
     }
     return artifacts.build();
-  }
-
-  private static LocalFileType getOutputType(@Nullable FileArtifactValue fileMetadata) {
-    if (fileMetadata == null) {
-      return LocalFileType.OUTPUT;
-    }
-    switch (fileMetadata.getType()) {
-      case DIRECTORY:
-        return LocalFileType.OUTPUT_DIRECTORY;
-      case SYMLINK:
-        return LocalFileType.OUTPUT_SYMLINK;
-      default:
-        return LocalFileType.OUTPUT_FILE;
-    }
   }
 
   @Override

--- a/src/test/java/com/google/devtools/build/lib/analysis/TargetCompleteEventTest.java
+++ b/src/test/java/com/google/devtools/build/lib/analysis/TargetCompleteEventTest.java
@@ -73,10 +73,9 @@ public class TargetCompleteEventTest extends AnalysisTestCase {
             artifactsToBuild.getAllArtifactsByOutputGroup(),
             /* announceTargetSummary= */ false);
 
-    assertThat(event.referencedLocalFiles()).hasSize(1);
-    LocalFile localFile = event.referencedLocalFiles().get(0);
-    assertThat(localFile.path).isEqualTo(artifact.getPath());
-    assertThat(localFile.type).isEqualTo(LocalFileType.OUTPUT_FILE);
+    assertThat(event.referencedLocalFiles())
+        .containsExactly(
+            new LocalFile(artifact.getPath(), LocalFileType.OUTPUT_FILE, artifact, metadata));
   }
 
   @Test
@@ -97,11 +96,9 @@ public class TargetCompleteEventTest extends AnalysisTestCase {
             artifactsToBuild.getAllArtifactsByOutputGroup(),
             /* announceTargetSummary= */ false);
 
-    assertThat(event.referencedLocalFiles()).hasSize(1);
-    LocalFile localFile = event.referencedLocalFiles().get(0);
-    assertThat(localFile.path).isEqualTo(artifact.getPath());
-    // TODO(tjgq): This should be reported as a directory.
-    assertThat(localFile.type).isEqualTo(LocalFileType.OUTPUT_FILE);
+    assertThat(event.referencedLocalFiles())
+        .containsExactly(
+            new LocalFile(artifact.getPath(), LocalFileType.OUTPUT_DIRECTORY, artifact, metadata));
   }
 
   @Test
@@ -110,10 +107,7 @@ public class TargetCompleteEventTest extends AnalysisTestCase {
         "defs.bzl",
         "def _impl(ctx):",
         "  d = ctx.actions.declare_directory(ctx.label.name)",
-        "  ctx.actions.run_shell(",
-        "    outputs = [d],",
-        "    command = 'touch %s/file.txt' % d.path,",
-        "  )",
+        "  ctx.actions.run_shell(outputs = [d], command = 'does not matter')",
         "  return DefaultInfo(files = depset([d]))",
         "dir = rule(_impl)");
     scratch.file(
@@ -123,16 +117,23 @@ public class TargetCompleteEventTest extends AnalysisTestCase {
         "filegroup(name = 'files', srcs = ['dir'])");
     ConfiguredTargetAndData ctAndData = getCtAndData("//:files");
     ArtifactsToBuild artifactsToBuild = getArtifactsToBuild(ctAndData);
-    SpecialArtifact artifact =
+    SpecialArtifact tree =
         (SpecialArtifact) Iterables.getOnlyElement(artifactsToBuild.getAllArtifacts().toList());
+    TreeFileArtifact fileChild =
+        TreeFileArtifact.createTreeOutput(tree, PathFragment.create("dir/file.txt"));
+    FileArtifactValue fileMetadata =
+        FileArtifactValue.createForNormalFile(new byte[] {1, 2, 3}, null, 10);
+    // A TreeFileArtifact can be a directory, when materialized by a symlink.
+    // See https://github.com/bazelbuild/bazel/issues/20418.
+    TreeFileArtifact dirChild = TreeFileArtifact.createTreeOutput(tree, PathFragment.create("sym"));
+    FileArtifactValue dirMetadata = FileArtifactValue.createForDirectoryWithMtime(123456789);
     TreeArtifactValue metadata =
-        TreeArtifactValue.newBuilder(artifact)
-            .putChild(
-                TreeFileArtifact.createTreeOutput(artifact, PathFragment.create("file")),
-                FileArtifactValue.createForNormalFile(new byte[] {1, 2, 3}, null, 10))
+        TreeArtifactValue.newBuilder(tree)
+            .putChild(fileChild, fileMetadata)
+            .putChild(dirChild, dirMetadata)
             .build();
     CompletionContext completionContext =
-        getCompletionContext(ImmutableMap.of(), ImmutableMap.of(artifact, metadata));
+        getCompletionContext(ImmutableMap.of(), ImmutableMap.of(tree, metadata));
 
     TargetCompleteEvent event =
         TargetCompleteEvent.successfulBuild(
@@ -141,11 +142,11 @@ public class TargetCompleteEventTest extends AnalysisTestCase {
             artifactsToBuild.getAllArtifactsByOutputGroup(),
             /* announceTargetSummary= */ false);
 
-    assertThat(event.referencedLocalFiles()).hasSize(1);
-    LocalFile localFile = event.referencedLocalFiles().get(0);
-    assertThat(localFile.path)
-        .isEqualTo(Iterables.getOnlyElement(metadata.getChildren()).getPath());
-    assertThat(localFile.type).isEqualTo(LocalFileType.OUTPUT_FILE);
+    assertThat(event.referencedLocalFiles())
+        .containsExactly(
+            new LocalFile(fileChild.getPath(), LocalFileType.OUTPUT_FILE, fileChild, fileMetadata),
+            new LocalFile(
+                dirChild.getPath(), LocalFileType.OUTPUT_DIRECTORY, dirChild, dirMetadata));
   }
 
   @Test
@@ -154,10 +155,7 @@ public class TargetCompleteEventTest extends AnalysisTestCase {
         "defs.bzl",
         "def _impl(ctx):",
         "  s = ctx.actions.declare_symlink(ctx.label.name)",
-        "  ctx.actions.symlink(",
-        "    output = s,",
-        "    target_path = '/some/path',",
-        "  )",
+        "  ctx.actions.symlink(output = s, target_path = 'does not matter')",
         "  return DefaultInfo(files = depset([s]))",
         "sym = rule(_impl)");
     scratch.file(
@@ -181,10 +179,9 @@ public class TargetCompleteEventTest extends AnalysisTestCase {
             artifactsToBuild.getAllArtifactsByOutputGroup(),
             /* announceTargetSummary= */ false);
 
-    assertThat(event.referencedLocalFiles()).hasSize(1);
-    LocalFile localFile = event.referencedLocalFiles().get(0);
-    assertThat(localFile.path).isEqualTo(artifact.getPath());
-    assertThat(localFile.type).isEqualTo(LocalFileType.OUTPUT_SYMLINK);
+    assertThat(event.referencedLocalFiles())
+        .containsExactly(
+            new LocalFile(artifact.getPath(), LocalFileType.OUTPUT_SYMLINK, artifact, metadata));
   }
 
   /** Regression test for b/165671166. */

